src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderRepository.cs

```diff
@@ -0,0 +1,378 @@
+                    {
+                        Row<byte[], ProvenBlockHeader> row =
+                            transaction.Select<byte[], ProvenBlockHeader>(ProvenBlockHeaderTable, i.ToBytes(false));
+
+                        if (row.Exists)
+                            items.Add(i, row.Value);
+                    }
+
+                    this.logger.LogTrace("(-)");
+
+                    return items;
+                }
+            });
+
+            return task;
+        }
+
+        /// <inheritdoc />
+        public Task<ProvenBlockHeader> GetAsync(int blockHeight)
+        {
+            Guard.NotNull(blockHeight, nameof(blockHeight));
+
+            Task<ProvenBlockHeader> task = Task.Run(() =>
+            {
+                this.logger.LogTrace("({0}:'{1}')", nameof(blockHeight), blockHeight);
+
+                using (DBreeze.Transactions.Transaction transaction = this.dbreeze.GetTransaction())
+                {
+                    ProvenBlockHeader header = null;
+
+                    transaction.SynchronizeTables(ProvenBlockHeaderTable);
+
+                    transaction.ValuesLazyLoadingIsOn = false;
+
+                    this.logger.LogTrace("Loading ProvenBlockHeader hash '{0}' from the database.", blockHeight);
+
+                    Row<byte[], ProvenBlockHeader> row =
+                        transaction.Select<byte[], ProvenBlockHeader>(ProvenBlockHeaderTable, blockHeight.ToBytes(false));
+
+                    if (row.Exists)
+                        header = row.Value;
+
+                    this.logger.LogTrace("(-)");
+
+                    return header;
+                }
+            });
+
+            return task;
+        }
+
+        /// <inheritdoc />
+        public Task PutAsync(List<ProvenBlockHeader> headers, HashHeightPair newTip)  
+        {
+            Guard.NotNull(headers, nameof(headers));
+            Guard.NotNull(newTip, nameof(newTip));
+
+            if ((this.provenBlockHeaderTip != null) && (newTip.Hash != this.provenBlockHeaderTip.HashPrevBlock))
+            {
+                this.logger.LogTrace("(-)[BLOCKHASH_MISMATCH]");
+                throw new InvalidOperationException("Invalid newTip block hash.");
+            }
+
+            Task task = Task.Run(() =>
+            {
+                this.logger.LogTrace("({0}.Count():{1})", nameof(headers), headers.Count());
+
+                using (DBreeze.Transactions.Transaction transaction = this.dbreeze.GetTransaction())
+                {
+                    transaction.SynchronizeTables(BlockHashHeightTable, ProvenBlockHeaderTable);
+
+                    this.InsertHeaders(transaction, headers, newTip);
+                    this.SetTip(transaction, newTip);
+
+                    transaction.Commit();
+                }
+
+                this.logger.LogTrace("(-)");
+            });
+
+            return task;
+        }
+
+        /// <inheritdoc />
+        public Task<HashHeightPair> GetTipHashHeightAsync()
```

Comment:
Should be private
And there should be a public method that returns cached variable that is a tip.

Reply:
This method now returns the cached variable and if it's null it then checks the disk.